File: server/src/engine/deck.ts
import { Card, Rank, Suit } from './types';
import { createCard, compareCardsDesc } from './card';

export function createDeck(): Card[] {
  const suits = [Suit.SPADE, Suit.HEART, Suit.CLUB, Suit.DIAMOND];
  const ranks = [
    Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT, Rank.NINE, Rank.TEN,
    Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE, Rank.TWO, Rank.THREE, Rank.FOUR,
  ];

  const deck: Card[] = [];
  for (const suit of suits) {
    for (const rank of ranks) {
      deck.push(createCard(suit, rank));
    }
  }

  deck.push(createCard(Suit.JOKER, Rank.SMALL_JOKER));
  deck.push(createCard(Suit.JOKER, Rank.BIG_JOKER));

  return deck;
}

export function shuffleDeck(deck: Card[]): Card[] {
  const shuffled = [...deck];
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
  }
  return shuffled;
}

export function deal(
  numPlayers: number,
  firstDealtPlayerIndex: number
): Card[][] {
  const deck = shuffleDeck(createDeck());
  const hands: Card[][] = Array.from({ length: numPlayers }, () => []);

  for (let i = 0; i < deck.length; i++) {
    const playerIdx = (firstDealtPlayerIndex + i) % numPlayers;
    hands[playerIdx].push(deck[i]);
  }

  for (const hand of hands) {
    hand.sort(compareCardsDesc);
  }

  return hands;
}

export function findRedFiveHolder(
  hands: Map<string, Card[]>
): string | null {
  for (const [playerId, hand] of hands) {
    const hasRedFive = hand.some(
      (c) => c.suit === Suit.HEART && c.rank === Rank.FIVE
    );
    if (hasRedFive) return playerId;
  }
  return null;
}

File: client/lib/engine/deck.dart
import 'dart:math';
import 'types.dart';
import 'card.dart';

List<Card> createDeck() {
  final suits = [Suit.S, Suit.H, Suit.C, Suit.D];
  final ranks = [
    Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT, Rank.NINE, Rank.TEN,
    Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE, Rank.TWO, Rank.THREE, Rank.FOUR,
  ];

  final deck = <Card>[];
  for (final suit in suits) {
    for (final rank in ranks) {
      deck.add(createCard(suit, rank));
    }
  }
  deck.add(createCard(Suit.JOKER, Rank.SMALL_JOKER));
  deck.add(createCard(Suit.JOKER, Rank.BIG_JOKER));
  return deck;
}

List<Card> shuffleDeck(List<Card> deck) {
  final shuffled = List<Card>.from(deck);
  final rng = Random();
  for (var i = shuffled.length - 1; i > 0; i--) {
    final j = rng.nextInt(i + 1);
    final temp = shuffled[i];
    shuffled[i] = shuffled[j];
    shuffled[j] = temp;
  }
  return shuffled;
}

List<List<Card>> deal(int numPlayers, int firstDealtPlayerIndex) {
  final deck = shuffleDeck(createDeck());
  final hands = List.generate(numPlayers, (_) => <Card>[]);

  for (var i = 0; i < deck.length; i++) {
    final playerIdx = (firstDealtPlayerIndex + i) % numPlayers;
    hands[playerIdx].add(deck[i]);
  }

  for (final hand in hands) {
    hand.sort(compareCardsDesc);
  }

  return hands;
}

String? findRedFiveHolder(Map<String, List<Card>> hands) {
  for (final entry in hands.entries) {
    final hasRedFive = entry.value.any((c) =>
        c.suit == Suit.H && c.rank == Rank.FIVE);
    if (hasRedFive) return entry.key;
  }
  return null;
}
